cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/PluginInstantiator.java

```diff
@@ -98,38 +112,86 @@ public ClassLoader getArtifactClassLoader(ArtifactDescriptor artifactDescriptor)
   }
 
   /**
-   * Loads and returns the {@link Class} of the given plugin class.
+   * Returns a {@link ClassLoader} for the given artifact identified by the given Location.
+   *
+   * @param location Location of the artifact
+   * @throws IOException if failed to expand the artifact jar to create the plugin ClassLoader
+   *
+   * @see PluginClassLoader
+   */
+  public ClassLoader getArtifactClassLoader(Location location) throws IOException {
+    try {
+      return locationClassLoaders.get(location);
+    } catch (ExecutionException e) {
+      Throwables.propagateIfInstanceOf(e.getCause(), IOException.class);
+      throw Throwables.propagate(e.getCause());
+    }
+  }
+
+  /**
+   * Loads and returns the {@link Class} of the given plugin.
    *
-   * @param artifactDescriptor descriptor for the artifact the plugin is from.
-   *                           It is used for creating the ClassLoader for the plugin.
-   * @param pluginClass information about the plugin class
+   * @param plugin {@link Plugin}
    * @param <T> Type of the plugin
    * @return the plugin Class
    * @throws IOException if failed to expand the plugin jar to create the plugin ClassLoader
    * @throws ClassNotFoundException if failed to load the given plugin class
    */
   @SuppressWarnings("unchecked")
-  public <T> Class<T> loadClass(ArtifactDescriptor artifactDescriptor,
+  public <T> Class<T> loadClass(Plugin plugin) throws IOException, ClassNotFoundException {
+    return (Class<T>) getArtifactClassLoader(plugin).loadClass(plugin.getPluginClass().getClassName());
+  }
+
+  /**
+   * Loads and returns the {@link Class} of the given plugin class.
+   *
+   * @param location location of the plugin
+   * @param pluginClass the plugin class
+   * @param <T> Type of the plugin
+   * @return the plugin class
+   * @throws IOException if failed to expand the plugin jar to create the plugin classloader
+   * @throws ClassNotFoundException if failed to load the given plugin class
+   */
+  @SuppressWarnings("unchecked")
+  public <T> Class<T> loadClass(Location location,
                                 PluginClass pluginClass) throws IOException, ClassNotFoundException {
-    return (Class<T>) getArtifactClassLoader(artifactDescriptor).loadClass(pluginClass.getClassName());
+    return (Class<T>) getArtifactClassLoader(location).loadClass(pluginClass.getClassName());
   }
 
   /**
    * Creates a new instance of the given plugin class.
```

Comment:
can you add some details to the javadocs about the pluginDir and expectations this class has for where the jars are? Without that its very hard for somebody who isn't already familiar with the class to figure out what's going on.